crates/core/src/topology.rs

```diff
@@ -39,49 +47,62 @@ const RANDOM_CLOSEST_DISTANCE: f64 = 1.0 / 1000.0;
 /// time window. The goal of this is to select the best connections over time
 /// from incoming join requests.
 pub(crate) struct TopologyManager {
-    slow_connection_evaluator: connection_evaluator::ConnectionEvaluator,
-    fast_connection_evaluator: connection_evaluator::ConnectionEvaluator,
-    request_density_tracker: request_density_tracker::RequestDensityTracker,
+    limits: Limits,
+    meter: Meter,
+    source_creation_times: DashMap<AttributionSource, Instant>,
+    slow_connection_evaluator: ConnectionEvaluator,
+    fast_connection_evaluator: ConnectionEvaluator,
+    request_density_tracker: RequestDensityTracker,
+    pub(crate) outbound_request_counter: OutboundRequestCounter,
     /// Must be updated when new neightbors are discovered.
     cached_density_map: CachedDensityMap,
-    pub this_peer_location: Location,
+    connection_acquisition_strategy: RwLock<ConnectionAcquisitionStrategy>,
 }
 
 impl TopologyManager {
     /// Create a new TopologyManager specifying the peer's own Location
-    pub(crate) fn new(this_peer_location: Location) -> Self {
+    pub(crate) fn new(limits: Limits) -> Self {
         TopologyManager {
-            slow_connection_evaluator: connection_evaluator::ConnectionEvaluator::new(
+            meter: Meter::new_with_window_size(100),
+            limits,
+            source_creation_times: DashMap::new(),
+            slow_connection_evaluator: ConnectionEvaluator::new(
                 SLOW_CONNECTION_EVALUATOR_WINDOW_DURATION,
             ),
-            fast_connection_evaluator: connection_evaluator::ConnectionEvaluator::new(
+            fast_connection_evaluator: ConnectionEvaluator::new(
                 FAST_CONNECTION_EVALUATOR_WINDOW_DURATION,
             ),
-            request_density_tracker: request_density_tracker::RequestDensityTracker::new(
+            request_density_tracker: RequestDensityTracker::new(
                 REQUEST_DENSITY_TRACKER_WINDOW_SIZE,
             ),
             cached_density_map: CachedDensityMap::new(),
-            this_peer_location,
+            outbound_request_counter: OutboundRequestCounter::new(
+                OUTBOUND_REQUEST_COUNTER_WINDOW_SIZE,
+            ),
+            connection_acquisition_strategy: RwLock::new(ConnectionAcquisitionStrategy::Fast),
         }
     }
 
     pub(crate) fn refresh_cache(
         &mut self,
-        current_neighbors: &BTreeMap<Location, usize>,
+        neighbor_locations: &BTreeMap<Location, Vec<Connection>>,
     ) -> Result<(), DensityMapError> {
         self.cached_density_map
-            .set(&self.request_density_tracker, current_neighbors)?;
+            .set(&self.request_density_tracker, neighbor_locations)?;
         Ok(())
     }
 
     /// Record a request and the location it's targeting
     pub(crate) fn record_request(
```

Comment:
I think now this and most functions under this module can be made non pub or pub(in crate::topology) encapsulating all the logic

Reply:
Doesn't `record_request()` need to be called from `Ring` - which isn't in the `topology` module?

Reply:
I think i was pointing at refresh_cache and other similar functions. But doesn't matter for now, will change it when we report resource usage so is easier to see what should be private etc.